File: tests/php/FileToolsTest.php
<?php

declare(strict_types=1);

namespace Tests\Php;

use AgentHarness\FileTools;
use AgentHarness\HasShell;
use AgentHarness\UsesTools;
use PHPUnit\Framework\TestCase;

final class _FileToolsAgent
{
    use HasShell;
    use UsesTools;

    public function __construct()
    {
        $this->initHasShell(cwd: '/work');
    }
}

final class FileToolsTest extends TestCase
{
    private function makeAgent(): _FileToolsAgent
    {
        return new _FileToolsAgent();
    }

    private function call(_FileToolsAgent $agent, string $name, array $args): mixed
    {
        return json_decode($agent->executeTool($name, $args), true);
    }

    // ---------- Read ----------

    public function testReadBasic(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.txt', "hello\nworld");
        FileTools::registerAll($a);
        $out = $this->call($a, 'Read', ['file_path' => '/work/a.txt']);
        $this->assertStringContainsString("     1\thello", $out);
        $this->assertStringContainsString("     2\tworld", $out);
    }

    public function testReadOffsetLimit(): void
    {
        $a = $this->makeAgent();
        $lines = [];
        for ($i = 1; $i <= 10; $i++) {
            $lines[] = "line{$i}";
        }
        $a->fs()->write('/work/big.txt', implode("\n", $lines));
        FileTools::registerAll($a);
        $out = $this->call($a, 'Read', ['file_path' => '/work/big.txt', 'offset' => 4, 'limit' => 2]);
        $this->assertStringContainsString("     4\tline4", $out);
        $this->assertStringContainsString("     5\tline5", $out);
        $this->assertStringNotContainsString('line3', $out);
        $this->assertStringNotContainsString('line6', $out);
    }

    public function testReadMissing(): void
    {
        $a = $this->makeAgent();
        FileTools::registerAll($a);
        $out = $this->call($a, 'Read', ['file_path' => '/nope.txt']);
        $this->assertArrayHasKey('error', $out);
        $this->assertStringContainsString('does not exist', $out['error']);
    }

    public function testReadDirectory(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/dir/a.txt', 'x');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Read', ['file_path' => '/work/dir']);
        $this->assertStringContainsString('directory', $out['error']);
    }

    public function testReadBinary(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/blob.bin', "text\x00stuff");
        FileTools::registerAll($a);
        $out = $this->call($a, 'Read', ['file_path' => '/work/blob.bin']);
        $this->assertStringContainsString('binary', $out['error']);
    }

    // ---------- Write ----------

    public function testWriteCreates(): void
    {
        $a = $this->makeAgent();
        FileTools::registerAll($a);
        $out = $this->call($a, 'Write', ['file_path' => '/work/new.txt', 'content' => 'hi']);
        $this->assertStringContainsString('written', $out);
        $this->assertSame('hi', $a->fs()->read('/work/new.txt'));
    }

    public function testWriteOverwrites(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/x.txt', 'old');
        FileTools::registerAll($a);
        $this->call($a, 'Write', ['file_path' => '/work/x.txt', 'content' => 'new']);
        $this->assertSame('new', $a->fs()->read('/work/x.txt'));
    }

    // ---------- Edit ----------

    public function testEditUnique(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'alpha beta gamma');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'beta', 'new_string' => 'BETA',
        ]);
        $this->assertStringContainsString('edited', $out);
        $this->assertSame('alpha BETA gamma', $a->fs()->read('/work/c.txt'));
    }

    public function testEditMissingString(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'hello');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'zzz', 'new_string' => 'qqq',
        ]);
        $this->assertStringContainsString('not found', $out['error']);
    }

    public function testEditMultipleNoReplaceAll(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'x x x');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'x', 'new_string' => 'y',
        ]);
        $this->assertStringContainsString('Found 3 matches', $out['error']);
    }

    public function testEditReplaceAll(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'x x x');
        FileTools::registerAll($a);
        $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'x', 'new_string' => 'y', 'replace_all' => true,
        ]);
        $this->assertSame('y y y', $a->fs()->read('/work/c.txt'));
    }

    public function testEditReadGateOffByDefault(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'hello');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'hello', 'new_string' => 'bye',
        ]);
        $this->assertStringContainsString('edited', $out);
    }

    public function testEditReadGateBlocks(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'hello');
        FileTools::registerAll($a, enforceReadGate: true);
        $out = $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'hello', 'new_string' => 'bye',
        ]);
        $this->assertStringContainsString('not been read', $out['error']);
    }

    public function testEditReadGatePassesAfterRead(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/c.txt', 'hello');
        FileTools::registerAll($a, enforceReadGate: true);
        $this->call($a, 'Read', ['file_path' => '/work/c.txt']);
        $out = $this->call($a, 'Edit', [
            'file_path' => '/work/c.txt', 'old_string' => 'hello', 'new_string' => 'bye',
        ]);
        $this->assertStringContainsString('edited', $out);
        $this->assertSame('bye', $a->fs()->read('/work/c.txt'));
    }

    // ---------- Glob ----------

    public function testGlobBasic(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', 'x');
        $a->fs()->write('/work/b.php', 'y');
        $a->fs()->write('/work/c.txt', 'z');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Glob', ['pattern' => '*.php']);
        sort($out);
        $this->assertSame(['/work/a.php', '/work/b.php'], $out);
    }

    public function testGlobRecursive(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', 'x');
        $a->fs()->write('/work/sub/b.php', 'y');
        $a->fs()->write('/work/sub/deeper/c.php', 'z');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Glob', ['pattern' => '**/*.php']);
        sort($out);
        $this->assertSame(['/work/a.php', '/work/sub/b.php', '/work/sub/deeper/c.php'], $out);
    }

    public function testGlobMtimeSort(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/old.php', 'x');
        $a->fs()->write('/work/middle.php', 'y');
        $a->fs()->write('/work/new.php', 'z');
        $a->fs()->write('/work/new.php', 'z2'); // bump mtime
        FileTools::registerAll($a);
        $out = $this->call($a, 'Glob', ['pattern' => '*.php']);
        $this->assertSame('/work/new.php', $out[0]);
    }

    public function testGlobPathDefaultsToCwd(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', 'x');
        $a->fs()->write('/elsewhere/b.php', 'y');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Glob', ['pattern' => '*.php']);
        $this->assertSame(['/work/a.php'], $out);
    }

    public function testGlobExplicitPath(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', 'x');
        $a->fs()->write('/elsewhere/b.php', 'y');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Glob', ['pattern' => '*.php', 'path' => '/elsewhere']);
        $this->assertSame(['/elsewhere/b.php'], $out);
    }

    // ---------- Grep ----------

    public function testGrepFilesWithMatches(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', "<?php\necho 'hi';");
        $a->fs()->write('/work/b.php', "echo 'hi';");
        $a->fs()->write('/work/c.php', 'x = 1');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => '^echo']);
        sort($out);
        $this->assertSame(['/work/a.php', '/work/b.php'], $out);
    }

    public function testGrepCount(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', "echo\necho\necho");
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => 'echo', 'output_mode' => 'count']);
        $this->assertSame([['path' => '/work/a.php', 'count' => 3]], $out);
    }

    public function testGrepContent(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', "alpha\nbeta TARGET here\ngamma");
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', [
            'pattern' => 'TARGET',
            'output_mode' => 'content',
            'line_numbers' => true,
            'context_before' => 1,
            'context_after' => 1,
        ]);
        $this->assertSame('/work/a.php', $out[0]['path']);
        $this->assertSame(2, $out[0]['matches'][0]['line']);
        $this->assertCount(3, $out[0]['matches'][0]['context']);
        $this->assertTrue($out[0]['matches'][0]['context'][1]['match']);
        $this->assertSame('beta TARGET here', $out[0]['matches'][0]['context'][1]['text']);
    }

    public function testGrepGlobFilter(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', 'echo');
        $a->fs()->write('/work/a.txt', 'echo');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => 'echo', 'glob' => '*.php']);
        $this->assertSame(['/work/a.php'], $out);
    }

    public function testGrepTypeFilter(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.php', 'echo');
        $a->fs()->write('/work/a.go', 'echo');
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => 'echo', 'type' => 'go']);
        $this->assertSame(['/work/a.go'], $out);
    }

    public function testGrepCaseInsensitive(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.txt', 'Hello WORLD');
        FileTools::registerAll($a);
        $this->assertSame([], $this->call($a, 'Grep', ['pattern' => 'world']));
        $this->assertSame(['/work/a.txt'], $this->call($a, 'Grep', ['pattern' => 'world', 'case_insensitive' => true]));
    }

    public function testGrepMultiline(): void
    {
        $a = $this->makeAgent();
        $a->fs()->write('/work/a.txt', "alpha\nbeta\ngamma");
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => 'alpha.*gamma', 'multiline' => true]);
        $this->assertSame(['/work/a.txt'], $out);
    }

    public function testGrepHeadLimit(): void
    {
        $a = $this->makeAgent();
        for ($i = 0; $i < 5; $i++) {
            $a->fs()->write("/work/f{$i}.txt", 'match');
        }
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => 'match', 'head_limit' => 2]);
        $this->assertCount(2, $out);
    }

    public function testGrepUnknownType(): void
    {
        $a = $this->makeAgent();
        FileTools::registerAll($a);
        $out = $this->call($a, 'Grep', ['pattern' => 'x', 'type' => 'bogus']);
        $this->assertStringContainsString('Unknown file type', $out['error']);
    }
}

File: src/go/filetools/filetools_test.go
package filetools

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"testing"

	"agent-harness/go/shell/vfs"
	"agent-harness/go/tools"
)

func newRig(t *testing.T, opts ...Option) (*tools.Registry, *vfs.VirtualFS, *FileTools) {
	t.Helper()
	v := vfs.New(nil)
	d := &vfs.BuiltinFilesystemDriver{FS: v}
	reg := tools.New()
	cwd := func() string { return "/work" }
	ft := RegisterAll(reg, d, cwd, opts...)
	return reg, v, ft
}

func mustExec(t *testing.T, reg *tools.Registry, name string, args map[string]any) any {
	t.Helper()
	raw, err := json.Marshal(args)
	if err != nil {
		t.Fatalf("marshal args: %v", err)
	}
	out, err := reg.Execute(context.Background(), name, raw)
	if err != nil {
		t.Fatalf("execute %s: %v", name, err)
	}
	return out
}

func mustExecErr(t *testing.T, reg *tools.Registry, name string, args map[string]any) string {
	t.Helper()
	raw, _ := json.Marshal(args)
	_, err := reg.Execute(context.Background(), name, raw)
	if err == nil {
		t.Fatalf("expected error from %s, got nil", name)
	}
	return err.Error()
}

// ---------- Read ----------

func TestRead_Basic(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.txt", "hello\nworld")
	out := mustExec(t, reg, "Read", map[string]any{"file_path": "/work/a.txt"}).(string)
	if !strings.Contains(out, "     1\thello") || !strings.Contains(out, "     2\tworld") {
		t.Fatalf("unexpected output: %q", out)
	}
}

func TestRead_OffsetLimit(t *testing.T) {
	reg, v, _ := newRig(t)
	var sb strings.Builder
	for i := 1; i <= 10; i++ {
		if i > 1 {
			sb.WriteByte('\n')
		}
		fmt.Fprintf(&sb, "line%d", i)
	}
	_ = v.WriteString("/work/big.txt", sb.String())
	out := mustExec(t, reg, "Read", map[string]any{"file_path": "/work/big.txt", "offset": 4, "limit": 2}).(string)
	if !strings.Contains(out, "     4\tline4") || !strings.Contains(out, "     5\tline5") {
		t.Fatalf("unexpected output: %q", out)
	}
	if strings.Contains(out, "line3") || strings.Contains(out, "line6") {
		t.Fatalf("output bled outside window: %q", out)
	}
}

func TestRead_Missing(t *testing.T) {
	reg, _, _ := newRig(t)
	msg := mustExecErr(t, reg, "Read", map[string]any{"file_path": "/nope.txt"})
	if !strings.Contains(msg, "does not exist") {
		t.Fatalf("expected does-not-exist error, got %q", msg)
	}
}

func TestRead_Directory(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/dir/a.txt", "x")
	msg := mustExecErr(t, reg, "Read", map[string]any{"file_path": "/work/dir"})
	if !strings.Contains(msg, "directory") {
		t.Fatalf("expected directory error, got %q", msg)
	}
}

func TestRead_Binary(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.Write("/work/blob.bin", []byte("text\x00stuff"))
	msg := mustExecErr(t, reg, "Read", map[string]any{"file_path": "/work/blob.bin"})
	if !strings.Contains(msg, "binary") {
		t.Fatalf("expected binary error, got %q", msg)
	}
}

// ---------- Write ----------

func TestWrite_Creates(t *testing.T) {
	reg, v, _ := newRig(t)
	out := mustExec(t, reg, "Write", map[string]any{"file_path": "/work/new.txt", "content": "hi"}).(string)
	if !strings.Contains(out, "written") {
		t.Fatalf("unexpected: %q", out)
	}
	got, _ := v.ReadString("/work/new.txt")
	if got != "hi" {
		t.Fatalf("content mismatch: %q", got)
	}
}

func TestWrite_Overwrites(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/x.txt", "old")
	mustExec(t, reg, "Write", map[string]any{"file_path": "/work/x.txt", "content": "new"})
	got, _ := v.ReadString("/work/x.txt")
	if got != "new" {
		t.Fatalf("content mismatch: %q", got)
	}
}

// ---------- Edit ----------

func TestEdit_Unique(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/c.txt", "alpha beta gamma")
	out := mustExec(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "beta", "new_string": "BETA",
	}).(string)
	if !strings.Contains(out, "edited") {
		t.Fatalf("unexpected: %q", out)
	}
	got, _ := v.ReadString("/work/c.txt")
	if got != "alpha BETA gamma" {
		t.Fatalf("content mismatch: %q", got)
	}
}

func TestEdit_Missing(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/c.txt", "hello")
	msg := mustExecErr(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "zzz", "new_string": "qqq",
	})
	if !strings.Contains(msg, "not found") {
		t.Fatalf("expected not-found, got %q", msg)
	}
}

func TestEdit_Multiple_NoReplaceAll(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/c.txt", "x x x")
	msg := mustExecErr(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "x", "new_string": "y",
	})
	if !strings.Contains(msg, "Found 3 matches") {
		t.Fatalf("expected 3-matches, got %q", msg)
	}
}

func TestEdit_ReplaceAll(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/c.txt", "x x x")
	mustExec(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "x", "new_string": "y", "replace_all": true,
	})
	got, _ := v.ReadString("/work/c.txt")
	if got != "y y y" {
		t.Fatalf("content mismatch: %q", got)
	}
}

func TestEdit_ReadGate_Off(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/c.txt", "hello")
	out := mustExec(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "hello", "new_string": "bye",
	}).(string)
	if !strings.Contains(out, "edited") {
		t.Fatalf("unexpected: %q", out)
	}
}

func TestEdit_ReadGate_Blocks(t *testing.T) {
	reg, v, _ := newRig(t, EnforceReadGate())
	_ = v.WriteString("/work/c.txt", "hello")
	msg := mustExecErr(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "hello", "new_string": "bye",
	})
	if !strings.Contains(msg, "not been read") {
		t.Fatalf("expected read-gate error, got %q", msg)
	}
}

func TestEdit_ReadGate_PassesAfterRead(t *testing.T) {
	reg, v, _ := newRig(t, EnforceReadGate())
	_ = v.WriteString("/work/c.txt", "hello")
	mustExec(t, reg, "Read", map[string]any{"file_path": "/work/c.txt"})
	out := mustExec(t, reg, "Edit", map[string]any{
		"file_path": "/work/c.txt", "old_string": "hello", "new_string": "bye",
	}).(string)
	if !strings.Contains(out, "edited") {
		t.Fatalf("unexpected: %q", out)
	}
	got, _ := v.ReadString("/work/c.txt")
	if got != "bye" {
		t.Fatalf("content mismatch: %q", got)
	}
}

// ---------- Glob ----------

func TestGlob_Basic(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "x")
	_ = v.WriteString("/work/b.go", "y")
	_ = v.WriteString("/work/c.txt", "z")
	out := mustExec(t, reg, "Glob", map[string]any{"pattern": "*.go"}).([]string)
	sort.Strings(out)
	if !equalStr(out, []string{"/work/a.go", "/work/b.go"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGlob_Recursive(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "x")
	_ = v.WriteString("/work/sub/b.go", "y")
	_ = v.WriteString("/work/sub/deeper/c.go", "z")
	out := mustExec(t, reg, "Glob", map[string]any{"pattern": "**/*.go"}).([]string)
	sort.Strings(out)
	want := []string{"/work/a.go", "/work/sub/b.go", "/work/sub/deeper/c.go"}
	if !equalStr(out, want) {
		t.Fatalf("got %v, want %v", out, want)
	}
}

func TestGlob_MtimeSort(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/old.go", "x")
	_ = v.WriteString("/work/middle.go", "y")
	_ = v.WriteString("/work/new.go", "z")
	_ = v.WriteString("/work/new.go", "z2") // bump mtime
	out := mustExec(t, reg, "Glob", map[string]any{"pattern": "*.go"}).([]string)
	if out[0] != "/work/new.go" {
		t.Fatalf("expected new.go first, got %v", out)
	}
}

func TestGlob_PathDefaultsToCwd(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "x")
	_ = v.WriteString("/elsewhere/b.go", "y")
	out := mustExec(t, reg, "Glob", map[string]any{"pattern": "*.go"}).([]string)
	if !equalStr(out, []string{"/work/a.go"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGlob_ExplicitPath(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "x")
	_ = v.WriteString("/elsewhere/b.go", "y")
	out := mustExec(t, reg, "Glob", map[string]any{"pattern": "*.go", "path": "/elsewhere"}).([]string)
	if !equalStr(out, []string{"/elsewhere/b.go"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

// ---------- Grep ----------

func TestGrep_FilesWithMatches(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "package x\nfunc Foo() {}")
	_ = v.WriteString("/work/b.go", "package y\nfunc Bar() {}")
	_ = v.WriteString("/work/c.go", "var z int")
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "^func"}).([]string)
	sort.Strings(out)
	if !equalStr(out, []string{"/work/a.go", "/work/b.go"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGrep_Count(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "func\nfunc\nfunc")
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "func", "output_mode": "count"})
	if !strings.Contains(fmt.Sprintf("%v", out), "/work/a.go") || !strings.Contains(fmt.Sprintf("%v", out), "3") {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGrep_Content(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "alpha\nbeta TARGET here\ngamma")
	raw, _ := json.Marshal(map[string]any{
		"pattern":        "TARGET",
		"output_mode":    "content",
		"line_numbers":   true,
		"context_before": 1,
		"context_after":  1,
	})
	out, err := reg.Execute(context.Background(), "Grep", raw)
	if err != nil {
		t.Fatalf("execute: %v", err)
	}
	// Marshal/unmarshal to assert against generic types.
	b, _ := json.Marshal(out)
	var got []map[string]any
	_ = json.Unmarshal(b, &got)
	if got[0]["path"] != "/work/a.go" {
		t.Fatalf("wrong path: %v", got[0])
	}
	matches := got[0]["matches"].([]any)
	first := matches[0].(map[string]any)
	if int(first["line"].(float64)) != 2 {
		t.Fatalf("wrong line: %v", first["line"])
	}
	ctx := first["context"].([]any)
	if len(ctx) != 3 {
		t.Fatalf("expected 3 context lines, got %d", len(ctx))
	}
	mid := ctx[1].(map[string]any)
	if mid["match"] != true || mid["text"] != "beta TARGET here" {
		t.Fatalf("unexpected mid: %v", mid)
	}
}

func TestGrep_GlobFilter(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "func")
	_ = v.WriteString("/work/a.txt", "func")
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "func", "glob": "*.go"}).([]string)
	if !equalStr(out, []string{"/work/a.go"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGrep_TypeFilter(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.go", "func")
	_ = v.WriteString("/work/a.py", "func")
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "func", "type": "py"}).([]string)
	if !equalStr(out, []string{"/work/a.py"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGrep_CaseInsensitive(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.txt", "Hello WORLD")
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "world"}).([]string)
	if len(out) != 0 {
		t.Fatalf("expected no matches, got %v", out)
	}
	out = mustExec(t, reg, "Grep", map[string]any{"pattern": "world", "case_insensitive": true}).([]string)
	if !equalStr(out, []string{"/work/a.txt"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGrep_Multiline(t *testing.T) {
	reg, v, _ := newRig(t)
	_ = v.WriteString("/work/a.txt", "alpha\nbeta\ngamma")
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "alpha.*gamma", "multiline": true}).([]string)
	if !equalStr(out, []string{"/work/a.txt"}) {
		t.Fatalf("unexpected: %v", out)
	}
}

func TestGrep_HeadLimit(t *testing.T) {
	reg, v, _ := newRig(t)
	for i := 0; i < 5; i++ {
		_ = v.WriteString(fmt.Sprintf("/work/f%d.txt", i), "match")
	}
	out := mustExec(t, reg, "Grep", map[string]any{"pattern": "match", "head_limit": 2}).([]string)
	if len(out) != 2 {
		t.Fatalf("expected 2, got %v", out)
	}
}

func TestGrep_UnknownType(t *testing.T) {
	reg, _, _ := newRig(t)
	msg := mustExecErr(t, reg, "Grep", map[string]any{"pattern": "x", "type": "bogus"})
	if !strings.Contains(msg, "Unknown file type") {
		t.Fatalf("expected unknown-type error, got %q", msg)
	}
}

// ---------- helpers ----------

func equalStr(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
